### PR TITLE
feat(FN-419): Add ACBS function to Bicep

### PR DESCRIPTION
**Pull request by OxonianCambion**
### Introduction

We need to be able to manage the infrastructure in declarative IaC - we've chosen Bicep.
The first phase is to be able to:

- produce a complete new deployment environment (feature) mirroring the current ones.
- be in a position to take over the extant environments.
- Note that there have been some manual changes to the environments. This work also functions as an audit of the current infrastructure.

Possible enhancements / updates will be noted but not implemented in the first phase.

This card covers:

- Creating the Azure Function for the ACBS function, wired up to the correct docker image from CI

### Resolution

The code:
- creates the Azure Function
- wired up with dummy environment variables
- with Application Insights
Note:
- that there isn't currently an image to deploy!
- It does not wire up the A Records for the private link yet. I was planning on just adding them all in a batch for all resources. @IgnatG I'm happy to just do it per-resource / PR if you prefer.
- I don't currently send in the environment variables to the module. This will need fixing - either now or when @IgnatG adds the secrets themselves.

### CODE REVIEW

File: infrastructure/resource_group_level/main.bicep [hunk 1]
```diff
@@ -1,4 +1,5 @@
 param location string  = resourceGroup().location
+// Expected values are 'feature', 'dev', 'test', 'staging' & 'prod'
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/main.bicep:2]: I think there is confusion between `qa`, `test` & `staging`. I think they are all the same, right?

File: infrastructure/resource_group_level/modules/function-acbs.bicep [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  MULESOFT_API_KEY: 'test-value'
+  MULESOFT_API_SECRET: 'test-value'
+  MULESOFT_API_UKEF_TF_EA_URL: 'test-value'
+  APIM_MDM_KEY: 'test-value'
+  APIM_MDM_URL: 'test-value'
+  APIM_MDM_VALUE: 'test-value' // different in staging and dev
+}
+
+// These values are taken from an export of Configuration on Dev
+// TODO:FN-419 Add to GitHub secrets?
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-acbs.bicep:22]: I see that in https://github.com/UK-Export-Finance/dtfs2/pull/1767 these values will change. @abhi-markan 

I can either wait for your PR to be merged, merge that into my long-lived, branch, rebase this, and make the change now, or make the change in a future card / PR, with a TODO here. I'll do the latter by default!

File: infrastructure/resource_group_level/modules/function-acbs.bicep [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  name: storageAccountName
+}
+
+var storageAccountKey = storageAccount.listKeys().keys[0].value
+
+// These values are hardcoded in the CLI scripts
+var settings = {
+  FUNCTIONS_WORKER_RUNTIME: 'node'
+  // TODO:FN-419 consider making this a parameter
+  WEBSITE_DNS_SERVER: '168.63.129.16'
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-acbs.bicep:47]: I can't see where this IP comes from. Any ideas?

File: infrastructure/resource_group_level/modules/function-acbs.bicep [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  WEBSITE_VNET_ROUTE_ALL: '1'
+}
+
+// These values are taken from an export of Configuration on Dev
+var additionalSettings = {
+  APPINSIGHTS_INSTRUMENTATIONKEY: applicationInsights.properties.InstrumentationKey
+  // TODO:FN-419 - Exported string had "EndpointSuffix=core.windows.net" in it. Check if needed
+  AzureWebJobsStorage: 'DefaultEndpointsProtocol=https;AccountName=${storageAccountName};AccountKey=${storageAccountKey}'
+  // TODO:FN-419 I think DOCKER_CUSTOM_IMAGE_NAME is overridden by linuxFxVersion
+  DOCKER_CUSTOM_IMAGE_NAME: dockerImageName
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-acbs.bicep:57]: By default, I'll just leave this as is.

File: infrastructure/resource_group_level/modules/function-acbs.bicep [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  FUNCTIONS_WORKER_RUNTIME: 'node'
+  // TODO:FN-419 consider making this a parameter
+  WEBSITE_DNS_SERVER: '168.63.129.16'
+  WEBSITE_VNET_ROUTE_ALL: '1'
+}
+
+// These values are taken from an export of Configuration on Dev
+var additionalSettings = {
+  APPINSIGHTS_INSTRUMENTATIONKEY: applicationInsights.properties.InstrumentationKey
+  // TODO:FN-419 - Exported string had "EndpointSuffix=core.windows.net" in it. Check if needed
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-acbs.bicep:54]: The EndpointSuffix isn't needed, right? https://learn.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string#create-a-connection-string-with-an-endpoint-suffix

File: infrastructure/resource_group_level/modules/function-acbs.bicep [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  AzureWebJobsStorage: 'DefaultEndpointsProtocol=https;AccountName=${storageAccountName};AccountKey=${storageAccountKey}'
+  // TODO:FN-419 I think DOCKER_CUSTOM_IMAGE_NAME is overridden by linuxFxVersion
+  DOCKER_CUSTOM_IMAGE_NAME: dockerImageName
+  DOCKER_ENABLE_CI: 'true'
+  DOCKER_REGISTRY_SERVER_URL: '${containerRegistryName}.azurecr.io'
+  DOCKER_REGISTRY_SERVER_USERNAME: dockerRegistryServerUsername
+  FUNCTION_APP_EDIT_MODE: 'readOnly'
+  FUNCTIONS_EXTENSION_VERSION: '~3'
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-acbs.bicep:62]: We can upgrade to V4 later. We should do it soon after though. V3 is EOL.

File: infrastructure/resource_group_level/modules/function-acbs.bicep [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+  location: location
+  tags: {
+    Environment: 'Preproduction'
+  }
+  kind: 'functionapp,linux,container'
+  properties: {
+    httpsOnly: false
+    serverFarmId: appServicePlanId
+    siteConfig: {
+      // TODO:FN-419 These siteConfig values are included in the much more comprehensive config object, produced by the export.
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-acbs.bicep:92]: Note that when I re-export this, I get the duplication, so I think all is well and I put the settings here, and just use the `functionAcbsAppSettings` resource below for environment variables.

File: infrastructure/resource_group_level/modules/function-acbs.bicep [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+// These values are hardcoded in the CLI scripts
+var settings = {
+  FUNCTIONS_WORKER_RUNTIME: 'node'
+  // TODO:FN-419 consider making this a parameter
+  WEBSITE_DNS_SERVER: '168.63.129.16'
+  WEBSITE_VNET_ROUTE_ALL: '1'
+}
+
+// These values are taken from an export of Configuration on Dev
+var additionalSettings = {
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-acbs.bicep:52]: I'm not totally happy with the Azure convention of `param`, `var`, `resource` ordering here as this var has a dependency on the applicationInsights resource below. I suppose it is ok.

File: infrastructure/resource_group_level/modules/function-acbs.bicep [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+  name: applicationInsightsName
+  location: location
+  kind: 'web'
+  properties: {
+    Application_Type: 'web'
+  }
+}
+
+
+// TODO:FN-419 Add automatic A Record generation.
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-acbs.bicep:172]: say if I should do this as part of the current PR, or later in a big batch.

File: infrastructure/resource_group_level/main.bicep (hunk 1 above)
Comment (IgnatG) [reply infrastructure/resource_group_level/main.bicep:2]: `staging` and `test` are the same. - they got merged into `staging` in the end

File: infrastructure/resource_group_level/modules/function-acbs.bicep (hunk 3 above)
Comment (IgnatG) [reply infrastructure/resource_group_level/modules/function-acbs.bicep:47]: good point. no idea :(

File: infrastructure/resource_group_level/modules/function-acbs.bicep (hunk 5 above)
Comment (IgnatG) [reply infrastructure/resource_group_level/modules/function-acbs.bicep:54]: double check if it's needed. if not, then we can remove it

File: infrastructure/resource_group_level/modules/function-acbs.bicep (hunk 6 above)
Comment (IgnatG) [reply infrastructure/resource_group_level/modules/function-acbs.bicep:62]: agreed

File: infrastructure/resource_group_level/modules/function-acbs.bicep [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+      minimumElasticInstanceCount: 1
+      // The following Fields have been added after comparing with dev
+      // TODO:FN-419 Check these values are ok.
+      vnetRouteAllEnabled: true
+      // Note that the following appear in the separate config object
+      ftpsState: 'Disabled'
+      scmMinTlsVersion: '1.0'
+      remoteDebuggingVersion: 'VS2019'
+      httpLoggingEnabled: true // false in staging
+      // TODO:FN-419 Note that the following appear in dev but not staging or prod
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-acbs.bicep:109]: I've moved this to a "mopping up things that will be easy to see when we can deploy" card.

**Review by IgnatG — APPROVED** [review]
LGTM

**Pull request merged**